### PR TITLE
gh-104745: Limit starting a patcher more than once without stopping it

### PR DESCRIPTION
**Pull request by Red4Ru**
Previously, this would cause an `AttributeError` if the patch stopped more than once after this, and would also disrupt the original patched object.

* Issue: gh-104745

### CODE REVIEW

File: Lib/unittest/mock.py [hunk 1]
```diff
@@ -1600,17 +1627,20 @@ def __enter__(self):
 
         new_attr = new
 
-        self.temp_original = original
-        self.is_local = local
-        self._exit_stack = contextlib.ExitStack()
+        self._started_context = _PatchStartedContext(
```
Comment (picnixz) [inline Lib/unittest/mock.py]: I think the naming `self._context` would be sufficient.

File: Lib/unittest/mock.py [hunk 2]
```diff
@@ -1600,17 +1627,20 @@ def __enter__(self):
...
+            exit_stack=contextlib.ExitStack(),
+            is_local=is_local,
+            target=self.getter(),
+            temp_original=original,
```
Comment (picnixz) [inline Lib/unittest/mock.py]: This attribute can be called `original` since we are anyway wrapping it in a new object.

File: Lib/unittest/mock.py [hunk 3]
```diff
@@ -1320,6 +1320,14 @@ def _check_spec_arg_typos(kwargs_to_check):
             )
 
 
+class _PatchStartedContext(object):
```
Comment (picnixz) [inline Lib/unittest/mock.py]: To reduce the memory footprint, how about using a namedtuple? I would also call it only `class _PatchContext` because it does not really give more information to have the "Started" IMO.

File: Lib/unittest/mock.py [hunk 4]
```diff
@@ -1600,17 +1627,20 @@ def __enter__(self):
 
         new_attr = new
 
-        self.temp_original = original
-        self.is_local = local
-        self._exit_stack = contextlib.ExitStack()
+        self._started_context = _PatchStartedContext(
+            exit_stack=contextlib.ExitStack(),
+            is_local=is_local,
+            target=self.getter(),
```
Comment (picnixz) [inline Lib/unittest/mock.py]:
```suggestion
            target=target,
```

You may not necessarily want to re-call self.getter() which could cause a side-effect.

File: Lib/unittest/mock.py [hunk 5]
```diff
@@ -1600,17 +1627,20 @@ def __enter__(self):
...
+        )
         try:
             setattr(self.target, self.attribute, new_attr)
             if self.attribute_name is not None:
                 extra_args = {}
                 if self.new is DEFAULT:
                     extra_args[self.attribute_name] =  new
                 for patching in self.additional_patchers:
-                    arg = self._exit_stack.enter_context(patching)
+                    arg = self._started_context.exit_stack.enter_context(patching)
```
Comment (picnixz) [inline Lib/unittest/mock.py]:
```suggestion
                    arg = exit_stack.enter_context(patching)
```

File: Lib/unittest/mock.py (hunk 1 above)
Comment (picnixz) [inline Lib/unittest/mock.py]:
```suggestion
		exit_stack = contextlib.ExitStack()
        self._started_context = _PatchStartedContext(
```

File: Lib/unittest/mock.py [hunk 6]
```diff
@@ -1600,17 +1627,20 @@ def __enter__(self):
 
         new_attr = new
 
-        self.temp_original = original
-        self.is_local = local
-        self._exit_stack = contextlib.ExitStack()
+        self._started_context = _PatchStartedContext(
+            exit_stack=contextlib.ExitStack(),
```
Comment (picnixz) [inline Lib/unittest/mock.py]:
```suggestion
            exit_stack=exit_stack,
```

File: Lib/unittest/mock.py [hunk 7]
```diff
@@ -1469,13 +1478,31 @@ def get_original(self):
             )
         return original, local
 
+    @property
+    def is_started(self):
```
Comment (picnixz) [inline Lib/unittest/mock.py]: Previously, those were writable attributes and now it's no more the case. Could there be some code in the wild assuming so? (for instance pytest which makes quite hacky things, though I don't know if they do hacky things with this specific part of CPython).

File: Misc/NEWS.d/next/Library/2024-11-10-18-14-51.gh-issue-104745.zAa5Ke.rst [hunk 8]
```diff
@@ -0,0 +1,5 @@
+Limit starting a patcher (from :func:`unittest.mock.patch`,
+:func:`unittest.mock.patch.object` or :func:`unittest.mock.dict`) more than
+once without stopping it. Previously, this would cause an
+:exc:`AttributeError` if the patch stopped more than once after this, and
+would also disrupt the original patched object.
```
Comment (ZeroIntensity) [inline Misc/NEWS.d/next/Library/2024-11-10-18-14-51.gh-issue-104745.zAa5Ke.rst]: NEWS entries should be pretty short:
```suggestion
Limit starting a patcher (from :func:`unittest.mock.patch`,
:func:`unittest.mock.patch.object` or :func:`unittest.mock.patch.dict`) more than
once without stopping it.
```

File: Misc/NEWS.d/next/Library/2024-11-10-18-14-51.gh-issue-104745.zAa5Ke.rst [hunk 9]
```diff
@@ -0,0 +1,5 @@
+Limit starting a patcher (from :func:`unittest.mock.patch`,
+:func:`unittest.mock.patch.object` or :func:`unittest.mock.dict`) more than
```
Comment (picnixz) [inline Misc/NEWS.d/next/Library/2024-11-10-18-14-51.gh-issue-104745.zAa5Ke.rst]:
```suggestion
:func:`unittest.mock.patch.object` or :func:`unittest.mock.patch.dict`) more than
```

File: Misc/NEWS.d/next/Library/2024-11-10-18-14-51.gh-issue-104745.zAa5Ke.rst (hunk 8 above)
Comment (picnixz) [reply Misc/NEWS.d/next/Library/2024-11-10-18-14-51.gh-issue-104745.zAa5Ke.rst]: It's actually `unittest.mock.patch.dict`: https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch.dict

Comment (ZeroIntensity) [reply Misc/NEWS.d/next/Library/2024-11-10-18-14-51.gh-issue-104745.zAa5Ke.rst]: Oh! Thank you!

File: Lib/unittest/mock.py (hunk 4 above)
Comment (Red4Ru) [reply Lib/unittest/mock.py]: Oh, good one! I refactored a bit and overlooked this after some copy-pasting

Comment (Red4Ru) [reply Lib/unittest/mock.py]: Interesting that `self.getter()` is also called in `get_original` method. Should I also add optional argument in here..?

File: Lib/unittest/mock.py (hunk 7 above)
Comment (Red4Ru) [reply Lib/unittest/mock.py]: I guess so... I have committed property setters just in case. Will write tests if needed when we figure out what to do with `temp_original`: do we preserve it and somehow deprecate or anything else

File: Lib/unittest/mock.py [hunk 10]
```diff
@@ -1469,13 +1474,51 @@ def get_original(self):
...
+    def is_local(self, value):
+        self._context.is_local = value
+
+    @property
+    def original(self):
+        return self._context.original
+
+    @original.setter
+    def original(self, value):
+        self._context.original = value
```
Comment (picnixz) [inline Lib/unittest/mock.py]: Actually when I said to rename it to `original`, it was only for the `context`'s field, not for the patch object itself!

For the patch's object, I think we can keep `temp_original` (or maybe you wanted to have the same name for both the context dataclass and the patch object? if so, sorry for the back-and-forth)

File: Lib/unittest/mock.py (hunk 4 above)
Comment (picnixz) [reply Lib/unittest/mock.py]: Let's try to keep the same behaviour as before unless there is a need to change.

File: Lib/unittest/mock.py (hunk 10 above)
Comment (Red4Ru) [reply Lib/unittest/mock.py]: Agreed, let's keep `temp_original` in `_patch` while renaming it for `._context`

File: Lib/unittest/mock.py (hunk 7 above)
Comment (Red4Ru) [reply Lib/unittest/mock.py]: Fixed them. A bit ugly but anyway these setters exist not for an intended usecase but for backwards compatibility only

File: Lib/unittest/mock.py [hunk 11]
```diff
@@ -1469,13 +1474,58 @@ def get_original(self):
...
+        return self._context.original
+
+    @is_local.setter
+    def is_local(self, value):
+        self._context = _PatchContext(
+            exit_stack=self._context.exit_stack,
+            is_local=value,
+            original=self._context.original,
+            target=self._context.target,
+        )
```
Comment (picnixz) [inline Lib/unittest/mock.py]: Urgh, I forgot that you cannot change the value of namedtuples. Ok, my suggestion using namedtuples was wrong. To reduce memory footprint, we can use `__slots__` in a regular class instead like you had before. That way, we save an `from collections import namedtuple` as well and simplify the property's setter. WDYT? (again sorry for this bad suggestion).

**Review by cjw296 — COMMENTED** [review]
Can you explain why we need the whole `_PatchContext` thing? 
What couldn't be achieved by just having a `self.is_started` boolean?

**Comment by Red4Ru** [comment]
> Can you explain why we need the whole `_PatchContext` thing? What couldn't be achieved by just having a `self.is_started` boolean?

Consistency, at the first place. Why add another field for that while we already have all the bunch of others which exist only when `_patch` is started? (They literally were deleted in `__exit__`)

**Review by cjw296 — CHANGES_REQUESTED** [review]
This feels like we could just add an `is_started` attribute, which is set to `True` when the patcher is started and `False` when it is stopped.

I'd prefer to see that smaller, simpler change, tbh. 

But, great to see the new tests in this area!

**Comment by Red4Ru** [comment]
> This feels like we could just add an `is_started` attribute, which is set to `True` when the patcher is started and `False` when it is stopped.
> 
> I'd prefer to see that smaller, simpler change, tbh.

I think that putting these fields in `_context` and calculating `is_started` based on it makes the code easier to maintain
What are the advantages of leaving it as it is by adding the `is_started` field to `_patch`?

**Comment by Red4Ru** [comment]
> What you propose introduces more complexity than is required to fix the bug, and to a piece of code that is already very complex.
> 
> The change I think you should make amounts to 3 lines, if I understand correctly, and I would prefer that change to the one currently in this PR.

Well, if we should minimize impact on existing code, I agree with this suggestion
But if someone wants to refactor this later, penultimate commit might be useful

File: Lib/test/test_unittest/testmock/testpatch.py [hunk 12]
```diff
@@ -1098,7 +1146,7 @@ def test_new_callable_patch(self):
 
         self.assertIsNot(m1, m2)
         for mock in m1, m2:
-            self.assertNotCallable(m1)
+            self.assertNotCallable(mock)
```
Comment (cjw296) [inline Lib/test/test_unittest/testmock/testpatch.py:1149]: Nice spot!

File: Lib/unittest/mock.py [hunk 13]
```diff
@@ -1491,7 +1494,7 @@ def __enter__(self):
             spec_set not in (True, None)):
             raise TypeError("Can't provide explicit spec_set *and* spec or autospec")
 
-        original, local = self.get_original()
+        original, is_local = self.get_original()
```
Comment (cjw296) [inline Lib/unittest/mock.py]: Please undo this change.

File: Lib/unittest/mock.py [hunk 14]
```diff
@@ -1601,8 +1604,9 @@ def __enter__(self):
         new_attr = new
 
         self.temp_original = original
-        self.is_local = local
+        self.is_local = is_local
```
Comment (cjw296) [inline Lib/unittest/mock.py]: Please undo this change.

File: Lib/unittest/mock.py [hunk 15]
```diff
@@ -1633,11 +1640,12 @@ def __exit__(self, *exc_info):
...
 
-        del self.temp_original
-        del self.is_local
-        del self.target
+        self.is_started = False
         exit_stack = self._exit_stack
         del self._exit_stack
+        del self.is_local
+        del self.temp_original
+        del self.target
```
Comment (cjw296) [inline Lib/unittest/mock.py:1646]: Please undo this change: `self.is_started = False` should be just before the return, and that feels like it should be the only change here.

Comment (Red4Ru) [reply Lib/unittest/mock.py:1646]: Undone all 3

**Review by cjw296 — APPROVED** [review]

**Comment by cjw296** [comment]
Turns out this PR did not have sufficient test coverage :-/ 

https://app.circleci.com/pipelines/github/testing-cabal/mock/2056/workflows/9168b46d-8368-4983-bc3f-d689b6bb2f2d/jobs/29375

**Comment by cjw296** [comment]
Turns out it was a bug in a test, fixed in https://github.com/python/cpython/pull/130787
